libfwupdplugin/fu-usb-hid-descriptor.c

```diff
@@ -0,0 +1,208 @@
+	return self->iface_number;
+}
+
+/**
+ * fu_usb_hid_descriptor_set_iface_number:
+ * @self: a #FuUsbHidDescriptor
+ *
+ * Sets the hid descriptor interface number.
+ *
+ * Since: 2.0.2
+ **/
+void
+fu_usb_hid_descriptor_set_iface_number(FuUsbHidDescriptor *self, guint8 iface_number)
+{
+	g_return_if_fail(FU_IS_USB_HID_DESCRIPTOR(self));
+	self->iface_number = iface_number;
+}
+
+/**
+ * fu_usb_hid_descriptor_get_descriptor_length:
+ * @self: a #FuUsbHidDescriptor
+ *
+ * Gets the HID descriptor length.
+ *
+ * Return value: integer
+ *
+ * Since: 2.0.2
+ **/
+gsize
+fu_usb_hid_descriptor_get_descriptor_length(FuUsbHidDescriptor *self)
+{
+	g_return_val_if_fail(FU_IS_USB_HID_DESCRIPTOR(self), 0);
+	return self->descriptor_length;
+}
+
+/**
+ * fu_usb_hid_descriptor_get_blob:
+ * @self: a #FuUsbHidDescriptor
+ *
+ * Gets the HID descriptor binary blob.
+ *
+ * Return value: (transfer none): The descriptor data
+ *
+ * Since: 2.0.2
+ **/
+GBytes *
+fu_usb_hid_descriptor_get_blob(FuUsbHidDescriptor *self)
+{
+	g_return_val_if_fail(FU_IS_USB_HID_DESCRIPTOR(self), NULL);
+	return self->blob;
+}
+
+/**
+ * fu_usb_hid_descriptor_set_blob:
+ * @self: a #FuUsbHidDescriptor
+ *
+ * Sets the HID descriptor binary blob.
+ *
+ * Since: 2.0.2
+ **/
+void
+fu_usb_hid_descriptor_set_blob(FuUsbHidDescriptor *self, GBytes *blob)
+{
+	g_return_if_fail(FU_IS_USB_HID_DESCRIPTOR(self));
+	if (self->blob != NULL)
+		g_bytes_unref(self->blob);
+	self->blob = g_bytes_ref(blob);
+}
+
+static gboolean
+fu_usb_hid_descriptor_parse(FuFirmware *firmware,
+			    GInputStream *stream,
+			    FwupdInstallFlags flags,
+			    GError **error)
+{
+	FuUsbHidDescriptor *self = FU_USB_HID_DESCRIPTOR(firmware);
+	g_autoptr(FuUsbHidDescriptorHdr) st = NULL;
+
+	/* parse */
+	st = fu_usb_hid_descriptor_hdr_parse_stream(stream, 0x0, error);
+	if (st == NULL)
+		return FALSE;
+	self->descriptor_length = fu_usb_hid_descriptor_hdr_get_class_descriptor_length(st);
+
+	/* success */
+	return TRUE;
+}
+
+static void
+fu_usb_hid_descriptor_codec_iface_init(FwupdCodecInterface *iface)
+{
+	iface->add_json = fu_usb_hid_descriptor_add_json;
+	iface->from_json = fu_usb_hid_descriptor_from_json;
+}
+
+static void
+fu_usb_hid_descriptor_init(FuUsbHidDescriptor *self)
+{
+}
+
+static void
+fu_usb_hid_descriptor_finalize(GObject *object)
+{
+	FuUsbHidDescriptor *self = FU_USB_HID_DESCRIPTOR(object);
+	if (self->blob != NULL)
+		g_bytes_unref(self->blob);
+	G_OBJECT_CLASS(fu_usb_hid_descriptor_parent_class)->finalize(object);
+}
+
+static void
+fu_usb_hid_descriptor_class_init(FuUsbHidDescriptorClass *klass)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS(klass);
+	FuFirmwareClass *firmware_class = FU_FIRMWARE_CLASS(klass);
+	object_class->finalize = fu_usb_hid_descriptor_finalize;
+	firmware_class->parse = fu_usb_hid_descriptor_parse;
+}
+
+/**
+ * fu_usb_hid_descriptor_new:
+ *
+ * Return value: a new #FuUsbHidDescriptor object.
+ *
+ * Since: 2.0.0
```

Comment:
2.0.2

Reply:
An entire days work for that comment! :)

Reply:
Heh, I didn't finish looking through the rest yet.  Juggling a bunch, I'll let you know if anything else stands out ;P